File: act col/last/ActCol07/example4.cpp
// =================================================================
//
// File: example4.cpp
// Authors:
// Description: This file contains the code to count the number of
//				even numbers within an array using Intel's TBB.
//              To compile: g++ example4.cpp -ltbb
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <iostream>
#include <iomanip>
#include <cmath>
#include <tbb/parallel_reduce.h>
#include <tbb/blocked_range.h>
#include "utils.h"

#define SIZE 1000000000 //1e9

using namespace std;
using namespace tbb;

using namespace std;
using namespace tbb;

class Even {
private:
	int *A, size,result;

public:
	Even(int *a, int s) : A(a), size(s),result(0) {}
    Even(Even &obj,split) : A(obj.A), size(obj.size),result(0) {}
    
	double getResult() const {
		return result;
	}

    void operator() (const blocked_range<int> &r){
		for (int i = r.begin(); i != r.end(); i++) {
			if((A[i]%2)==0){
				result +=1;
			}
		}
    }


    void join(const Even &x) {
		result += x.result;
	}
};

int main(int argc, char* argv[]) {
	int *a;
	double ms;
    int result;

	a = new int[SIZE];
	fill_array(a, SIZE);
	display_array("a", a);

	cout << "Starting..." << endl;
	ms = 0;
	
	for (int i = 0; i < N; i++) {
		start_timer();

		// call your method here.
		//obj.doTask();
        Even obj(a,SIZE);
        parallel_reduce(blocked_range<int>(0, SIZE), obj);
        result = obj.getResult();
		ms += stop_timer();
	}
	cout << "result = "<< setprecision(15) << result<<endl;
	cout << "avg time = " << setprecision(15) << (ms / N) << " ms" << endl;

	delete [] a;
	return 0;
}

File: last/last/ActCol04/Example4.java
// =================================================================
//
// File: Example4.java
// Authors: Martin Noboa - A01704052
// 		   Bernardo Estrada - A01704320
// Description: This file contains the code to count the number of
//				even numbers within an array. The time this implementation
//				takes will be used as the basis to calculate the
//				improvement obtained with parallel technologies.
//
//			This implementation uses the Thread class.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

// ======Outputs====================================================
// Single Thread
// sum = 941896832
// avg time = 34.2 ms
//
// Multi Thread
// sum = 941896832
// avg time = 16.6 ms
//
// Speedup = 2.06x

public class Example4 extends Thread {
	private static final int SIZE = 100_000_000;
	private int array[];
	private int result;
	private int start, end;

	public Example4(int array[], int start, int end) {
		this.array = array;
		this.result = 0;
		this.start = start;
		this.end = end;
	}

	public int getResult() {
		return result;
	}

	public void calculate() {
		result = 0;
		for (int i = start; i < end; i++){
			if (array[i] % 2 == 0){
				result += array[i];
			}
		}
	}

	public void run() {
		calculate();
	}

	public static void main(String args[]) {
		int array[] = new int[SIZE];
		long startTime, stopTime;
		double acum = 0;
		int res = 0;

		Utils.fillArray(array);
		Utils.displayArray("array", array);

		Example4 threads[] = new Example4[Utils.MAXTHREADS];
		int blockSize = SIZE / Utils.MAXTHREADS;

		acum = 0;
		System.out.printf("Starting with %d threads\n", Utils.MAXTHREADS);
		for (int i = 0; i < Utils.N; i++) {
			res = 0;
			startTime = System.currentTimeMillis();

			for (int j = 0; j < threads.length; j++) {
				threads[j] = new Example4(array, j * blockSize, (j + 1) * blockSize);
				threads[j].start();
			}
			try {
				for (int j = 0; j < threads.length; j++) {
					threads[j].join();
					res += threads[j].getResult();
				}
			} catch (InterruptedException ie) {
				ie.printStackTrace();
			}

			stopTime = System.currentTimeMillis();

			acum += (stopTime - startTime);
		}
		System.out.printf("sum = %d\n", res);
		System.out.printf("avg time = %.5f ms\n", (acum / Utils.N));
	}
}

